sql/test/test.go

```diff
@@ -0,0 +1,83 @@
+package test
+
+import (
+	"path/filepath"
+	"slices"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+type Database interface {
+	sql.Executor
+	Close() error
+}
+
+type DBFuncs[DB Database] struct {
+	Schema   func() (*sql.Schema, error)
+	Open     func(uri string, opts ...sql.Opt) (DB, error)
+	InMemory func(opts ...sql.Opt) DB
+}
```

Comment:
Why not use an interface instead? It would require creating an implementation forwarding to the functions but it would seem less hacky.

```suggestion
type DBFuncs[DB Database] interface {
	Schema(*sql.Schema, error)
	Open(uri string, opts ...sql.Opt) (DB, error)
	InMemory(opts ...sql.Opt) DB
}
```

Reply:
Well, that'll cost some extra tens of lines, but probably should be more correct indeed, will fix

Reply:
At the moment the tests say nothing. For instance which package is tested here and what happens in those tests?

```go
var fns = test.DBFuncs[*Database]{Schema: Schema, Open: Open, InMemory: InMemory}

func TestDatabase_1(t *testing.T) {
	test.ExecuteFirstTest(t, fns)
}

func TestDatabase_2(t *testing.T) {
	test.ExecuteSecondTest(t, fns)
}
```

Reply:
When inlining this generic test into the actual packages it's only ~20 lines of code more but its much clearer what each test does:

```bash
$ git diff --stat 
 sql/localsql/localsql_test.go | 63 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----
 sql/statesql/statesql_test.go | 63 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----
 sql/test/test.go              | 83 -----------------------------------------------------------------------------------
 3 files changed, 116 insertions(+), 93 deletions(-)
```

Reply:
It's a bit of tradeoff b/c while I agree that inlining the tests makes them more readable here, yet it also makes it more likely for them to drift apart over time as `localsql` / `statesql` pkgs are updated, e.g. to include some coded migrations. The `TestSchema` test is especially undesirable to be "forked" like this, as it provides means of updating the schema should it drift from what you get after running all the migrations on an empty database. Maybe I could keep the test bodies imported, but add some clarifying comments?

Reply:
But with the same argument the `statesql` and `localsql` packages can drift apart. They are basically identical in their code - why not implement those in a generic way?

I know DRY should be avoided, but in cases like here I think drying up the code is actually an anti-pattern - it makes the code harder to read and debug and if `statesql` and `localsql` are changed in the future the generic test most likely becomes more and more complex to accommodate for this instead of just having two simple tests.

Reply:
Removed the "golden" test, replacing it with `make generate` / `make test-generate`, and inlined the idempotent schema test in the `localsql` and `statesql` packages